trlx/trainer/nn/ilql_models.py

```diff
@@ -331,3 +335,144 @@ def dummy_inputs(self):
+            past_key_values=past_key_values,
+            encoder_outputs=encoder_outputs,
+            output_attentions=output_attentions,
+            output_hidden_states=output_hidden_states,
+        )
+        out = self.base_model(**forward_kwargs)
+
+        hs = out.decoder_hidden_states[-1]
+
+        logits = self.base_model.lm_head(hs)
+        qs, target_qs, vs = self.ilql_heads(hs, states_ixs=states_ixs, actions_ixs=actions_ixs)
+        encoder_outputs = (out.encoder_last_hidden_state, out.encoder_hidden_states, out.encoder_attentions)
+        return logits, qs, target_qs, vs, out.past_key_values, encoder_outputs
+
+    def generate(
+        self,
+        input_ids,
+        attention_mask=None,
+        decoder_input_ids=None,
+        past_key_values=None,
+        encoder_outputs=None,
+        beta=1,
+        max_new_tokens=32,
+        max_length=1024,
+        temperature=1,
+        top_k=20,
+        logit_mask=None,
+        pad_token_id=None,
+        eos_token_id=None,
+    ):
+        """
+        Generates samples akin to hf's `.generate` but with custom logp prepossessing:
+        changing token probabilities as to how advantageous they would be
+        according to value functions estimations.
+        """
+        if attention_mask is None:
+            attention_mask = input_ids.not_equal(pad_token_id)
+        samples = input_ids.clone()
+        max_new_tokens = min(max_new_tokens, max_length - input_ids.shape[1])
+        if decoder_input_ids is None:
+            decoder_input_ids = input_ids.new_zeros(input_ids.shape[0], 1, dtype=torch.long, device=input_ids.device)
+        finished = torch.zeros(input_ids.shape[0], 1, dtype=torch.long, device=input_ids.device)
+        for _ in range(max_new_tokens):
+            out = self.forward(
+                input_ids=input_ids,
+                attention_mask=attention_mask,
+                decoder_input_ids=decoder_input_ids[:, -1].unsqueeze(-1),
+                past_key_values=past_key_values,
+                encoder_outputs=encoder_outputs,
+            )
+            logits, _, target_qs, vs, past_key_values, encoder_outputs = out
+            if self.ilql_config.two_qs:
+                qs = torch.minimum(target_qs[0][:, -1, :], target_qs[1][:, -1, :])
+            else:
+                qs = target_qs[:, -1, :]
+
+            logits = logits[:, -1, :]
+            vs = vs[:, -1, :]
+            adv = qs - vs
+            pi_beta = F.log_softmax(logits, -1)
+            pi_top_k = topk_mask(pi_beta + beta * adv, top_k)
+            pi = F.softmax(pi_top_k / temperature, -1)
+            next_tokens = torch.multinomial(pi, num_samples=1)
+            if eos_token_id is not None:
+                next_tokens = (1 - finished) * next_tokens + finished * pad_token_id
+                finished = (next_tokens == eos_token_id).long()
+            else:
+                raise ValueError("eos_token_id must be provided")
```

Comment:
this check is great, however it's better to do it at the start of the function and maybe not in the loop